src/Nethermind/Nethermind.Blockchain/TransactionPools/TransactionPool.cs

```diff
@@ -0,0 +1,218 @@
+            NewPending?.Invoke(this, new TransactionEventArgs(transaction));
+            NotifyPeers(SelectPeers(GetAvailablePeers(transaction)), transaction);
+            FilterAndStoreTransaction(transaction, blockNumber);
+        }
+
+        private void FilterAndStoreTransaction(Transaction transaction, UInt256 blockNumber)
+        {
+            var filters = _filters.Values.ToArray();
+            for (var i = 0; i < filters.Length; i++)
+            {
+                var filter = filters[i];
+                if (!filter.CanAdd(transaction))
+                {
+                    return;
+                }
+            }
+
+            _transactionStorage.Add(transaction, blockNumber);
+            if (_logger.IsDebug)
+            {
+                _logger.Debug($"Added transaction: {transaction.Hash}");
+            }
+        }
+
+        public void DeleteTransaction(Keccak hash)
+        {
+            _pendingTransactions.TryRemove(hash, out var transaction);
+            var filters = _filters.Values.ToArray();
+            for (var i = 0; i < filters.Length; i++)
+            {
+                var filter = filters[i];
+                if (!filter.CanDelete(transaction))
+                {
+                    return;
+                }
+            }
+
+            _transactionStorage.Delete(hash);
+            if (_logger.IsDebug)
+            {
+                _logger.Debug($"Deleted transaction: {hash}");
+            }
+        }
+
+        public void AddReceipt(TransactionReceipt receipt)
+        {
+            _receiptStorage.Add(receipt);
+        }
+
+        public event EventHandler<TransactionEventArgs> NewPending;
+
+        private void NotifyPeers(ISynchronizationPeer[] peers, Transaction transaction)
+        {
+            if (peers.Length == 0)
+            {
+                return;
+            }
+
+            for (var i = 0; i < peers.Length; i++)
+            {
+                var peer = peers[i];
+                peer.SendNewTransaction(transaction);
+            }
+
+            if (_logger.IsDebug)
+            {
+                _logger.Debug($"Notified {peers.Length} about transaction: {transaction.Hash}");
+            }
+        }
+
+        private ISynchronizationPeer[] SelectPeers(ISynchronizationPeer[] availablePeers)
+        {
+            if (availablePeers.Length == 0)
+            {
+                return new ISynchronizationPeer[0];
+            }
+
+            var selectedPeers = new List<ISynchronizationPeer>();
+            for (var i = 0; i < availablePeers.Length; i++)
+            {
+                var peer = availablePeers[i];
+                if (_peerThreshold >= GetRandomNumber())
+                {
+                    selectedPeers.Add(peer);
+                }
+            }
+
+            return selectedPeers.ToArray();
+        }
+
+        private ISynchronizationPeer[] GetAvailablePeers(Transaction transaction)
+        {
```

Comment:
can we merge getavailable and select so we do not have to iterate twice and allocate lists? this will be inefficient with the amount of transactions that flow during heavy network congestion times

Reply:
Ok